### ISSUE TITLE
Major Project Data science

### ISSUE THREAD

**Issue by debalinachakraborty944-blip** [adam-p/markdown-here#984]
MAJOR PROJECT 
PROJECT 2
DEBALINA CHAKRABORTY 
DATA SCIENCE 

---

 1. Import & Explore the Data

Start by loading and inspecting:

import pandas as pd
import numpy as np

df = pd.read_csv("sensor-data.csv")

print(df.shape)        # (1567, 592)
print(df.head())
print(df.info())
print(df.describe())

Key checks:

Are all features numeric? (usually yes in sensor data)

Identify missing values:

df.isnull().sum().sort_values(ascending=False)

---

 2. Data Cleansing

 Missing Value Treatment

Sensor datasets often have many missing values.

Strategy:

Drop columns with too many missing values (e.g. >50%)

threshold = 0.5
df = df.loc[:, df.isnull().mean() < threshold]

Fill remaining missing values:

from sklearn.impute import SimpleImputer

imputer = SimpleImputer(strategy='median')
df_imputed = pd.DataFrame(imputer.fit_transform(df), columns=df.columns)

---

Drop Irrelevant Features

Remove constant or near-constant features:

from sklearn.feature_selection import VarianceThreshold

selector = VarianceThreshold(threshold=0.01)
df_var = selector.fit_transform(df_imputed)

---

 3. Data Analysis & Visualization

Univariate Analysis

import matplotlib.pyplot as plt
import seaborn as sns

df_imputed.hist(figsize=(15,10))

Insights:

Many sensor values may be skewed

Some features may have zero variance

---

 Target Distribution

sns.countplot(x=df_imputed['target'])

 Expect imbalance (more pass = -1)

---

 Correlation Analysis

corr = df_imputed.corr()
sns.heatmap(corr)

Helps detect redundant features

---

Multivariate Analysis

Use PCA for visualization:

from sklearn.decomposition import PCA

pca = PCA(n_components=2)
reduced = pca.fit_transform(df_imputed.drop('target', axis=1))

---

4. Data Preprocessing

Split Features & Target

X = df_imputed.drop('target', axis=1)
y = df_imputed['target']

---

Handle Imbalance (VERY IMPORTANT)

from imblearn.over_sampling import SMOTE

smote = SMOTE()
X_res, y_res = smote.fit_resample(X, y)

---

 Train-Test Split

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X_res, y_res, test_size=0.2, random_state=42
)

---

 Standardization

from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

---
5. Model Training, Testing & Tuning

We’ll use 3 models:

---

 Model 1: Random Forest

from sklearn.ensemble import RandomForestClassifier

rf = RandomForestClassifier()

from sklearn.model_selection import GridSearchCV

params = {
    'n_estimators': [100, 200],
    'max_depth': [5, 10, None]
}

grid_rf = GridSearchCV(rf, params, cv=5)
grid_rf.fit(X_train, y_train)

---

 Model 2: Support Vector Machine

from sklearn.svm import SVC

params = {
    'C': [0.1, 1, 10],
    'kernel': ['rbf', 'linear']
}

grid_svm = GridSearchCV(SVC(), params, cv=5)
grid_svm.fit(X_train, y_train)

---

Model 3: Naive Bayes

from sklearn.naive_bayes import GaussianNB

nb = GaussianNB()
nb.fit(X_train, y_train)

---
Evaluation

from sklearn.metrics import classification_report

models = {
    "Random Forest": grid_rf.best_estimator_,
    "SVM": grid_svm.best_estimator_,
    "Naive Bayes": nb
}

for name, model in models.items():
    y_pred = model.predict(X_test)
    print(f"\n{name}")
    print(classification_report(y_test, y_pred))

---

Dimensionality Reduction (Important!)

Use PCA:

pca = PCA(n_components=50)
X_train_pca = pca.fit_transform(X_train)
X_test_pca = pca.transform(X_test)

 Then retrain models → often improves performance

---

 Model Comparison

Model	Train Accuracy	Test Accuracy	Notes

Random Forest	High	High	Handles noise well
SVM	Medium	High	Good for high-dim data
Naive Bayes	Low	Medium	Fast but simple

 Usually Random Forest or SVM wins

Save Final Model

import joblib

best_model = grid_rf.best_estimator_
joblib.dump(best_model, "final_model.pkl")

---

6. Conclusion

Key Findings:

Not all 591 features are useful

Removing noise improves accuracy

SMOTE significantly improves minority class prediction

PCA helps reduce dimensionality without losing much info

Final Model Choice:

👉 Random Forest (or SVM if better results)
Because:

Handles high-dimensional data well

Robust to noise.
